### PR TITLE
Newpost tim/drools dataquality checking

### PR DESCRIPTION
**Pull request by timurgen**

### CODE REVIEW

File: _posts/2018-07-23-DroolsDataQualityChecking.md [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+layout: post
+title: Using Drools (Business Rules Management System) for data quality checking. 
+categories: Data analysis
+tags: [Drools, Data quality, Elasticsearch, Apache Kafka]
+author: TimurSamkharadze
+---
+
+# Introduction
+
+In SYSCO AS we deal with a lot of data both structured and unstructured, streamed and batch.  It comes from different sources and goes to different destinations, and in some situations we want to measure its quality.  Data format may evolve over time  and we need to have possibility to change rules dynamically without restarting working systems.
```
Comment (jeqo) [inline _posts/2018-07-23-DroolsDataQualityChecking.md]: nit: break into lines of 80-120 chars to be easier to read and comment.

File: _posts/2018-07-23-DroolsDataQualityChecking.md [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+# Tools 
+
+JDK 7+, Maven, preferred IDE or text editor installed.
+
+# Problem description
+
+A system containing data about electrical installations with a lot of data sources from different organizations. Data quality problems may occur often in such systems. Date formats, floating point delimiters, geodata representation may differ. Typo and incorrect values may occur from time to time. So, let’s gonna fix it!
+
+We wrote an little micro-service application that fetches data from Elasticsearch instance, processes it and puts into a Apache Kafka topic.
```
Comment (jeqo) [inline _posts/2018-07-23-DroolsDataQualityChecking.md]: typo: "a little"

File: _posts/2018-07-23-DroolsDataQualityChecking.md [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+```java
+	while (elasticService.hasNextBatch()) {
+		Collection<Component> processedBatch = processBatch(kContainer, elasticService.getNextBatch());
+
+		processedBatch.forEach((element) -> {
+			kafkaService.put(element.id, objectMapper.toJson(element));
+		});
+	}
+```
+Component is simple POJO containing information about processed piece of data and messages from rule engine.
```
Comment (jeqo) [inline _posts/2018-07-23-DroolsDataQualityChecking.md]: nit: space between code and paragraph

File: _posts/2018-07-23-DroolsDataQualityChecking.md [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    then
+        action
+end
+```
+
+Several rules may be placed into one rule file.
+
+
+### 1. Check component type.
+Every component we check should belong to one of two types "GENERIC" or "METER"
```
Comment (jeqo) [inline _posts/2018-07-23-DroolsDataQualityChecking.md]: nit: space between title and paragraph

File: _posts/2018-07-23-DroolsDataQualityChecking.md [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        )
+    then
+        $component.messages.add("ID invalid");
+end
+
+```
+
+# Conclusion
+
+In this article we covered only one little bit of functionality provided by Drools rule engine that is still enough to make rules for data validation and conversion. More complicated rules can do everything that usual Java code can do but with full separation from application logic. So, you don't need redeploy your application every time when rules change.
```
Comment (jeqo) [inline _posts/2018-07-23-DroolsDataQualityChecking.md]: What does the last sentence mean? You won't need to recompile only if your rules are an external ref right? That's not explicit.

Comment (jeqo) [reply _posts/2018-07-23-DroolsDataQualityChecking.md]: Also, I don't see where is the reference to the rule file.

**Review by jeqo — COMMENTED** [review]
@timurgen check comments, mostly small things. Do we have a public repo to ref and test this?

File: _posts/2018-07-23-DroolsDataQualityChecking.md (hunk 1 above)
Comment (timurgen) [reply _posts/2018-07-23-DroolsDataQualityChecking.md]: fixed

File: _posts/2018-07-23-DroolsDataQualityChecking.md (hunk 2 above)
Comment (timurgen) [reply _posts/2018-07-23-DroolsDataQualityChecking.md]: fixed

File: _posts/2018-07-23-DroolsDataQualityChecking.md (hunk 3 above)
Comment (timurgen) [reply _posts/2018-07-23-DroolsDataQualityChecking.md]: fixed

File: _posts/2018-07-23-DroolsDataQualityChecking.md (hunk 4 above)
Comment (timurgen) [reply _posts/2018-07-23-DroolsDataQualityChecking.md]: fixed

File: _posts/2018-07-23-DroolsDataQualityChecking.md (hunk 5 above)
Comment (timurgen) [reply _posts/2018-07-23-DroolsDataQualityChecking.md]: added code for rule loading, explained hot reload more explicit

**Comment by timurgen** [comment]
We don't have public repo but private https://github.com/sysco-middleware/infokomp-dataquality-scorer
we don't use kafka actually here, but elastic on both ends. I placed kafka to this article just to have more appropriate keywords

**Comment by jeqo** [comment]
@timurgen you think is posible to have a repo with a working sample (nothing big, just to probe the point)?

**Comment by timurgen** [comment]
@jeqo  I have one I used to test rules already. I can add it to Sysco repo or to my own repo and add link to article if needed.

**Comment by jeqo** [comment]
@timurgen great!! I'd say sysco org is better as it is referenced from a Sysco blog post

**Comment by timurgen** [comment]
@jeqo  fixed

**Comment by jeqo** [comment]
LGTM.

**Pull request merged**